src/com/amplitude/api/AmplitudeClient.java

```diff
@@ -1372,13 +1379,19 @@ public String getUserId() {
      * @param userId the user id
      * @return the AmplitudeClient
      */
-    public AmplitudeClient setUserId(String userId) {
+    public AmplitudeClient setUserId(final String userId) {
         if (!contextAndApiKeySet("setUserId()")) {
             return this;
         }
 
-        this.userId = userId;
-        dbHelper.insertOrReplaceKeyValue(USER_ID_KEY, userId);
+        final AmplitudeClient client = this;
+        runOnLogThread(new Runnable() {
+            @Override
+            public void run() {
+                client.userId = userId;
```

Comment:
It's done on the log thread to maintain ordering, say if you already have 2 log events queued up when you set a new user id, you want those 2 events to still get the old user id.